Introduce Azure Traffic Manager 2018-04-01 API version with new features

allencal opened the pull request:
This is a copy of https://github.com/Azure/azure-rest-api-specs/pull/3162/files with fixes to that PR (no auth to update original PR).

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process.

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

allencal committed:
New API 2018-04-01

New API 2018-04-01 with support for multi-value and custom subnets.

allencal committed:
Create Endpoint-DELETE-External.json

allencal committed:
Examples for new API vdersion.

allencal committed:
Fixes for custom header examples

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-node

AutorestCI commented:
# Automation for azure-sdk-for-java

AutorestCI commented:
# Automation for azure-sdk-for-go

dsgouda commented on specification/trafficmanager/resource-manager/Microsoft.Network/stable/2018-04-01/trafficmanager.json:
```diff
@@ -0,0 +1,1578 @@
+        "tags": [
+          "Endpoints"
+        ],
+        "operationId": "Endpoints_Update",
```
Please check if the patch/put/delete operations are long running and mark them accordingly

dsgouda replied:
You can leave it as is if the operation is not long running

dsgouda commented on specification/trafficmanager/resource-manager/Microsoft.Network/stable/2018-04-01/trafficmanager.json:
```diff
@@ -0,0 +1,1578 @@
+      }
+    },
+    "/providers/Microsoft.Network/checkTrafficManagerNameAvailability": {
+      "post": {
```
This service endpoint does need have a subscriptionId/resourceGroup in the path?

allencal replied:
It does not need a subscription ID. This is a globally available query to find out if a DNS name has been reserved.

dsgouda replied:
Thanks @allencal is this service already in production? If not we need an ARM\API review here

allencal replied:
This service is already in production. Some new operations are being added for custom headers, custom subnets and MultiValue response.

dsgouda commented on specification/trafficmanager/resource-manager/Microsoft.Network/stable/2018-04-01/trafficmanager.json:
```diff
@@ -0,0 +1,1578 @@
+            }
+          },
+          "default": {
+            "description": "Default response. It will be deserialized as per the Error definition.",
```
nit: it is good documentation to specify all the error codes returned by the service. An example [here](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/sql/resource-manager/Microsoft.Sql/stable/2015-05-01/capabilities.json#L49)

allencal replied:
This is an old operation that hasn't changed for years. Since there are no parameters to verify, there's actually no other expected response than the 200.

dsgouda replied:
@allencal you can disregard this

dsgouda commented on specification/trafficmanager/resource-manager/Microsoft.Network/stable/2018-04-01/trafficmanager.json:
```diff
@@ -0,0 +1,1578 @@
+{
```
Is this spec being referenced in a config file? There must be atleast one tag associated with this to generate the SDKs

allencal replied:
I don't understand this question. Can you clarify?

sergey-shandar replied:
@allencal is this file was referenced from `readme.md` file? All spec files has to be referenced from `readme.md` file. Otherwise, they will be ignored.

allencal replied:
Thanks. Added readme.md and readme.ruby.md.

dsgouda commented:
Blocking on ARM review to be completed here https://github.com/Azure/azure-rest-api-specs/pull/3162/files

dsgouda commented:
Talked to @allencal offline and looks like he is porting changes from https://github.com/Azure/azure-rest-api-specs/pull/3162 here along with some of his additional changes.
He will address @ravbhatnagar 's comments posted there and update this PR
@ravbhatnagar could you please finish the ARM review here

dsgouda commented:
Assigning this to @sergey-shandar since he has more context here

allencal reviewed (commented):
Per ravbhatnagar's comments in PR 3162, added /default after trafficManagerUserMetricsKeys.

allencal committed:
Update versions in readme.md and readme.ruby.md

Update versions in readme.md and readme.ruby.md to 2018-04-01

ravbhatnagar commented:
Looks good!

sergey-shandar commented on specification/trafficmanager/resource-manager/readme.ruby.md:
```diff
@@ -16,6 +17,9 @@ batch:
   - tag: package-2017-03
   - tag: package-2017-05
   - tag: package-2017-09-preview-only
+  - tag: package-2018-02
```
`readme.md` has no `package-2018-02` tag.

sergey-shandar commented on specification/trafficmanager/resource-manager/Microsoft.Network/stable/2018-04-01/trafficmanager.json:
```diff
@@ -0,0 +1,1578 @@
+      "properties": {
+        "id": {
+          "type": "string",
+          "description": "Fully qualified resource Id for the resource. Ex - /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/trafficManagerProfiles/{resourceName}"
```
`"readonly": true`

sergey-shandar commented on specification/trafficmanager/resource-manager/Microsoft.Network/stable/2018-04-01/trafficmanager.json:
```diff
@@ -0,0 +1,1578 @@
+        },
+        "name": {
+          "type": "string",
+          "description": "The name of the resource"
```
`"readonly": true`

sergey-shandar commented on specification/trafficmanager/resource-manager/Microsoft.Network/stable/2018-04-01/trafficmanager.json:
```diff
@@ -0,0 +1,1578 @@
+        },
+        "type": {
+          "type": "string",
+          "description": "The type of the resource. Ex- Microsoft.Network/trafficmanagerProfiles."
```
`"readonly": true`

allencal committed:
Removing package-2018-02 from Ruby

Linter is failing package-2018-02 for unclear reasons. Since there are no apparent errors to fix in readme.md, I'm removing the failing reference in this file.

allencal committed:
Add readonly property to id, name, type

Add readonly property to id, name, type

sergey-shandar reviewed (approved):
Merged with condition that `readme.ruby.md` will be fixed in the next PR soon. @sarangan12

sergey-shandar commented:
@dsgouda are you ok with the changes? Please, approved the PR. Otherwise I can't merge.

dsgouda commented:
LGTM

markjbrown commented:
@allencal @dsgouda @sergey-shandar 
This swagger spec is causing rendering issues in docs for User Metrics Keys and HeatMap.

https://review.docs.microsoft.com/en-us/rest/api/trafficmanager/trafficmanagerusermetricskeys?branch=trafficmanager 

Specifically, the "Try It" feature renders for the REST operations (it shouldn't) but not for the examples themselves. There are also issues with the Definitions for UserMetricsModel in multiple places.

Can you take a look and fix? When you merge the fix you should be able to return to this docs link above and see if it is fixed.

thanks.
